src/github.com/matrix-org/dendrite/clientapi/routing/password.go

```diff
@@ -0,0 +1,177 @@
+// previous parameters with the ones supplied. This mean you cannot "build up" request params.
+
+// a generic flowRequest type for any auth call to UIAA handler to be included in UIAA.go Todo referenced in (#413)
+type userInteractiveFlowRequest struct {
+	// parameters
+	Password string `json:"password"`
+	Username string `json:"username"`
+	Admin    bool   `json:"admin"`
+
+	// user-interactive auth params
+	Auth authDict `json:"auth"`
+
+	// Application Services place Type in the root of their registration
+	// request, whereas clients place it in the authDict struct.
+	Type authtypes.LoginType `json:"type"`
+}
+
+type changePasswordRequest struct {
+	// user-interactive auth params (necessary to include to make call to UIAA handler)
+	userInteractiveFlowRequest
+
+	// password change parameters
+	NewPassword string `json:"new_password"`
+}
+
+// ChangePassword implements a /password request.
+func ChangePassword(
+	req *http.Request,
+	accountDB *accounts.Database,
+	cfg *config.Dendrite,
+) util.JSONResponse {
+
+	// Todo username and password aren't in request as per spec. Use whoami referenced in #411
+	var r changePasswordRequest
+	resErr := httputil.UnmarshalJSONRequest(req, &r)
+	if resErr != nil {
+		return *resErr
+	}
+
+	// Retrieve or generate the sessionID
+	sessionID := r.Auth.Session
+	if sessionID == "" {
+		// Generate a new, random session ID
+		sessionID = util.RandomString(sessionIDLength)
+	}
+
+	// If no auth type is specified by the client, send back the list of available flows
+	if r.Auth.Type == "" {
+		return util.JSONResponse{
+			Code: 401,
+			// Todo replace by password.Flows when available
+			JSON: newUserInteractiveResponse(sessionID,
+				cfg.Derived.Registration.Flows, cfg.Derived.Registration.Params),
+		}
+	}
+
+	// Squash username to all lowercase letters
+	r.Username = strings.ToLower(r.Username)
+
+	// validate new password
+	if resErr = validatePassword(r.NewPassword); resErr != nil {
+		return *resErr
+	}
+
+	logger := util.GetLogger(req.Context())
+	logger.WithFields(log.Fields{
+		"username":   r.Username,
+		"auth.type":  r.Auth.Type,
+		"session_id": r.Auth.Session,
+	}).Info("Processing password request")
+
+	// Todo go through User Interactive Auth uncomment references in #413  and
+	// authflows defined in cfg.derieved like that for registration
+
+	//jsonRes := HandleUserInteractiveFlow(req, r.userInteractiveFlowRequest, sessionID, cfg,
+	//	userInteractiveResponse{
+	//		// passing the list of allowed Flows and Params
+	//		cfg.Derived.Password.Flows,
+	//		nil,
+	//		cfg.Derived.Password.Params,
+	//		"",
+	//	})
+	//if jsonRes.Code == 200 {
+	//	// cast JSON to userInteractiveHandlerResponse
+	//	res := jsonRes.JSON.(userInteractiveHandlerResponse)
+	//
+	//	return completeUpdation(
+	//		req.Context(),
+	//		accountDB,
+	//		r.Username,
+	//		r.Password,
+	//		res.AppserviceID,
+	//		r.InitialDisplayName)
+	//}
+	//return jsonRes
```

Comment:
I would say to not add code as a comment. Especially when you are done with implementing.

Reply:
Actually that code was to be added when #413 is merged or at least its motive is served. I wrote it as a draft for reviewers. Will remove it if u say.